### PR TITLE
Fix a couple of bugs on the instrumentation generator

### PR DESCRIPTION
**Pull request by vandonr**
## Summary of changes

Fixing some bugs I encountered when working on my latest integrations:
 - when instrumenting interfaces, the instrumentation kind was not set
 - integration name was generated using the method name, which is never what we want. I replaced it with a hypothetical integration ID bearing the name of the DLL, which (1) has chances of being correct and (2) will be red in the editor if it's not the right one (better than having everything compile and not work without knowing why)
 - do not suggest type constraints when the instrumented type is generic (see [limitations](https://github.com/DataDog/dd-trace-dotnet/blob/master/docs/development/AutomaticInstrumentation.md#current-limitations)). Here it'd be even nicer to generate DuckCopy structs instead of interfaces for the stuff we cannot constraint, but it looked complicated.
 - there is one kind of template type that was not handled, I don't know why or how it's different from the other, I just did a fix.

see commit list for full list

## Reason for change

generate usable code

## Test coverage

no tests AFAIK 😬 I used the tool on Avro, comparing with what I ended up putting in [my PR](https://github.com/DataDog/dd-trace-dotnet/pull/6882/files) because it was an instrumentation full of corner cases.

### CODE REVIEW

**Comment by andrewlock** [comment]
## Execution-Time Benchmarks Report :stopwatch:

Execution-time results for samples comparing the following branches/commits:
- [This PR (6985)](https://github.com/DataDog/dd-trace-dotnet/tree/0104231163cbc88fdcc4d2b55b1db3813a313f01)
- [master](https://github.com/DataDog/dd-trace-dotnet/tree/37820885329c2af69173b1a2df567e55082e6c2f)

Execution-time benchmarks measure the whole time it takes to execute a program. And are intended to measure the one-off costs. Cases where the execution time results for the PR are worse than latest master results are shown in **red**. The following thresholds were used for comparing the execution times:
* Welch test with statistical test for significance of **5%**
* Only results indicating a difference greater than **5%** and **5 ms** are considered.


Note that these results are based on a _single_ point-in-time result for each branch. For full results, see the [dashboard](https://ddstaging.datadoghq.com/dashboard/4qn-6fi-54p/apm-net-execution-time-benchmarks).

Graphs show the p99 interval based on the mean and StdDev of the test run, as well as the mean value of the run (shown as a diamond below the graph).
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Framework 4.8) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6985) - mean (75ms)  : 70, 80
     .   : milestone, 75,
    master - mean (68ms)  : 65, 70
     .   : milestone, 68,

    section CallTarget+Inlining+NGEN
    This PR (6985) - mean (1,061ms)  : 1012, 1109
     .   : milestone, 1061,
    master - mean (1,007ms)  : 982, 1031
     .   : milestone, 1007,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6985) - mean (107ms)  : 103, 110
     .   : milestone, 107,
    master - mean (102ms)  : 99, 104
     .   : milestone, 102,

    section CallTarget+Inlining+NGEN
    This PR (6985) - mean (722ms)  : 684, 760
     .   : milestone, 722,
    master - mean (691ms)  : 673, 710
     .   : milestone, 691,

```
```mermaid
gantt
    title Execution time (ms) FakeDbCommand (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6985) - mean (97ms)  : 92, 101
     .   : milestone, 97,
    master - mean (88ms)  : 86, 91
     .   : milestone, 88,

    section CallTarget+Inlining+NGEN
    This PR (6985) - mean (691ms)  : 655, 726
     .   : milestone, 691,
    master - mean (655ms)  : 634, 677
     .   : milestone, 655,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Framework 4.8) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6985) - mean (189ms)  : 185, 193
     .   : milestone, 189,
    master - mean (189ms)  : 186, 192
     .   : milestone, 189,

    section CallTarget+Inlining+NGEN
    This PR (6985) - mean (1,108ms)  : 1085, 1131
     .   : milestone, 1108,
    master - mean (1,111ms)  : 1077, 1145
     .   : milestone, 1111,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET Core 3.1) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6985) - mean (268ms)  : 263, 273
     .   : milestone, 268,
    master - mean (268ms)  : 263, 272
     .   : milestone, 268,

    section CallTarget+Inlining+NGEN
    This PR (6985) - mean (903ms)  : 801, 1006
     .   : milestone, 903,
    master - mean (876ms)  : 845, 907
     .   : milestone, 876,

```
```mermaid
gantt
    title Execution time (ms) HttpMessageHandler (.NET 6) 
    dateFormat  X
    axisFormat %s
    todayMarker off
    section Baseline
    This PR (6985) - mean (260ms)  : 257, 263
     .   : milestone, 260,
    master - mean (261ms)  : 257, 264
     .   : milestone, 261,

    section CallTarget+Inlining+NGEN
    This PR (6985) - mean (860ms)  : 828, 892
     .   : milestone, 860,
    master - mean (867ms)  : 836, 898
     .   : milestone, 867,

```

File: tracer/src/Datadog.AutoInstrumentation.Generator/EditorHelper.cs [hunk 1]
```diff
@@ -56,18 +56,23 @@ public static string CleanTypeName(string typeName)
...
+            return $"nameof(IntegrationId.{CleanTypeName(value)})";
+        }
+
+        return string.Empty;
     }
 
     public static string GetMinimumVersion(MethodDef methodDef)
     {
         var version = methodDef.DeclaringType.DefinitionAssembly.Version;
-        return $"{version.Major}.{version.Minor}.{version.Build}";
```
Comment (andrewlock) [inline tracer/src/Datadog.AutoInstrumentation.Generator/EditorHelper.cs:70]: I wonder if we _Shouldn't_ change this - this is _Safer_ unless you've tested on earlier versions, and it's much easier to make the edit to `.0.0` rather than the other way? 🤔

File: tracer/src/Datadog.AutoInstrumentation.Generator/Resources/Data/Integration.cs [hunk 2]
```diff
@@ -24,7 +24,7 @@ namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.$(Namespace);
     ParameterTypeNames = $(ParameterTypeNames),
     MinimumVersion = "$(MinimumVersion)",
     MaximumVersion = "$(MaximumVersion)",
-    IntegrationName = "$(IntegrationName)")]
+    IntegrationName = $(IntegrationValue)$(IntegrationKind))]
```
Comment (andrewlock) [inline tracer/src/Datadog.AutoInstrumentation.Generator/Resources/Data/Integration.cs:27]: Renaming `IntegrationName` to `IntegrationValue`, when it's used to set the `IntegrationName` seems an odd choice... but 🤷‍♂️  meh 😄

**Review by andrewlock — APPROVED** [review]
Thanks!

File: tracer/src/Datadog.AutoInstrumentation.Generator/EditorHelper.cs (hunk 1 above)
Comment (vandonr) [reply tracer/src/Datadog.AutoInstrumentation.Generator/EditorHelper.cs:70]: ah yes, my reasoning was "what are the chances you're inspecting exactly the version that starts to have the methods you want to instrument ?" and I concluded "pretty much none", while wanting to instrument from the first release of the major is pretty standard.
I also assumed we have tests in place that run on min and max, so if min is too low, we'd see it, right ? While if it's too high, it's kinda invisible ?

Comment (andrewlock) [reply tracer/src/Datadog.AutoInstrumentation.Generator/EditorHelper.cs:70]: > while wanting to instrument from the first release of the major is pretty standard.

It actually varies significantly with different integrations

> "what are the chances you're inspecting exactly the version that starts to have the methods you want to instrument ?" and I concluded "pretty much none"

I don't think we can safely make that conclusion tbh, in fact I normally explicitly do this, and the reason is:

> I also assumed we have tests in place that run on min and max, so if min is too low, we'd see it, right? 

Yeah, no, we don't. We typically have a floating "high" version we test on, and we test on other _specific_ versions.

Ultimately it's down to the person writing the integration to choose what _needs_ to be tested, and to choose which versions to instrument, but I would argue the change makes it _slightly_ more risky that we ship something untested, while also making it _slightly_ more likely that you're going to accidentally not support versions that you _could_. IMO, that's a better trade off, but I'm not passionate about it.

File: tracer/src/Datadog.AutoInstrumentation.Generator/Resources/Data/Integration.cs (hunk 2 above)
Comment (andrewlock) [reply tracer/src/Datadog.AutoInstrumentation.Generator/Resources/Data/Integration.cs:27]: I would personally have kept it as `IntegrationName`, but I repeat, meh 😄 A big fat meh.
